.adr/0017-missed-chain-events.md

```diff
@@ -0,0 +1,131 @@
+
+```go
+	// Prevent go routines from processing events before checkForMissedEvents completes
+	ecs.eventTracker.mu.Lock()
+	{
+		ecs.wg.Add(3)
+		go ecs.listenForEventLogs(errChan, eventSub, eventChan, eventQuery)
+		go ecs.listenForNewBlocks(errChan, newBlockSub, newBlockChan)
+		go ecs.listenForErrors(errChan)
+
+		// Search for any missed events emitted while this node was offline
+		err = ecs.checkForMissedEvents(startBlock)
+		if err != nil {
+			return nil, err
+		}
+	}
+	ecs.eventTracker.mu.Unlock()
+```
+
+The `listenForEventLogs` go routine also tries to acquire the `mutex` before updating the `eventTracker.events` queue but will be blocked until `checkForMissedEvents` completes. This means that `listenForEventLogs` will still be able to detect new chain events, but will not add them to the queue or trigger event processing until `checkForMissedEvents` finishes queueing old events.
+
+### [2] Put `Channel` data protection logic in `Channel` class
+
+Add logic to the `Channel` class to protect its data from stale updates rather than relying on an external component (`chainservice`) to only feed it sanitized data. This seems less error-prone and it could be useful to add a `Channel.LastUpdated` struct in case a user wants to know when the channel was last changed. The `chainservice` will still be expected to feed chain events in order, but the added `Channel` logic will act as another layer of protection against faulty data updates.
+
+The `Channel` class should keep track of the `TxIndex` associated with each chain event in case there are multiple events related to a single channel within the same block.
+
+```go
+type Channel struct {
+	state.FixedPart
+	Id      types.Destination
+	MyIndex uint
+
+	OnChain  OnChainData
+	OffChain OffChainData
+
+  LastChainUpdate ChainUpdateData
+}
+
+type ChainUpdateData struct {
+  BlockNum uint64
+  TxIndex  uint
+}
+
+func (c *Channel) isNewChainEvent(event chainservice.Event) bool {
+	return event.BlockNum() > c.LastChainUpdate.BlockNum ||
+		(event.BlockNum() == c.LastChainUpdate.BlockNum && event.TxIndex() > c.LastChainUpdate.TxIndex)
+}
+
+func (c *Channel) UpdateWithChainEvent(event chainservice.Event) (*Channel, error) {
+	if !c.isNewChainEvent(event) {
+		return nil, fmt.Errorf("chain event older than channel's last update")
+	}
+	// Process event
+	...
+
+	// Update Channel.LastChainUpdate
+	c.LastChainUpdate.BlockNum = event.BlockNum()
+	c.LastChainUpdate.TxIndex  = event.TxIndex()
+	return c, nil
+}
+```
+
+### [3] Updating `lastBlockNum`: `engine` periodically reads from `chainservice`
+
+Instead of the `chainservice` alerting the `engine`, the `engine` can request the block number from the `chainservice`. This means the `lastBlockNumSeen` wouldn't get updated every block, but that's not a major concern since the node
```

Comment:
This sentence trails off?

Reply:
Fixed in https://github.com/statechannels/go-nitro/pull/1624/commits/06b800ff68d376a29ee776e21e9b8a074fcd8ff6